bin/proving-service/src/utils.rs

```diff
@@ -143,3 +145,34 @@ pub async fn create_response_with_error_message(
     session.write_response_header(Box::new(header)).await?;
     Ok(true)
 }
+
+/// Checks if a port is available for use.
+///
+/// # Arguments
+/// * `host` - The host to bind to.
+/// * `port` - The port to check.
+/// * `port_name` - A descriptive name for the port (for logging purposes).
+///
+/// # Returns
+/// * `Ok(())` if the port is available.
+/// * `Err(ProvingServiceError::PortAlreadyInUse)` if the port is already in use.
+pub fn check_port_availability(
+    host: &str,
+    port: u16,
+    port_name: &str,
+) -> Result<(), ProvingServiceError> {
+    let addr = format!("{}:{}", host, port);
+    match TcpListener::bind(&addr) {
+        Ok(_) => {
+            // Port is available, we can proceed
+            tracing::info!("Port {} is available for {}", port, port_name);
+            Ok(())
+        },
+        Err(e) => {
+            // Port is already in use, log an error and return an error
+            let error_msg = format!("{} port {} is already in use: {}", port_name, port, e);
+            tracing::error!("{}", error_msg);
```

Comment:
Is this redundant? Seems like something similar to this should be surfaced to `main()` eventually since you are returning `Err` here, right?

Reply:
As an aside (not related to this PR), for converting from the error type to `String` (as done from outside of this function), it would probably be nice if the whole chain of errors could be included